### ISSUE TITLE
binutils use of rpath causes link errors on at least Ubuntu Xenial when libfl-dev is installed

### ISSUE THREAD

**Issue by akesandgren** [hpcugent/easybuild-easyconfigs#3409]
The easyblock for binutils is forcibly setting RPATH in the resulting binaries (ar, ranlib etc) to start with
/usr/lib:/lib:/usr/lib64:/lib64:/usr/lib/x86_64-linux-gnu/:/lib/x86_64-linux-gnu/

This is causing problems when libfl-dev is installed on Ubuntu Xenial (at least). That package contains
/usr/lib/x86_64-linux-gnu/libfl.so.2.0.0 which is not a real shared object but a ld script which just does
INPUT( /usr/lib/x86_64-linux-gnu/libfl_pic.a )

This results in several binutils binaries built with GCCcore to be unusable, due to the dependency on libfl.so.2, they simply fail to load with
ar: error while loading shared libraries: /usr/lib/x86_64-linux-gnu/libfl.so.2: invalid ELF header

A personal reflection is that one should NEVER add system lib-dirs to RPATH, it will only cause problems and it is not what RPATH is there for.

So, I suggest removing that part from the easyblock of binutils completely.
Or, at the very least add a
        gcccoreroot = get_software_root('GCCcore')
        if gcccoreroot:
            libs = ''
        else:
or similar around it (although that wouldn't help if someone decided to build it with something completely different...).

**Comment by boegel** [comment]
@wpoely86 thoughts?

**Comment by wpoely86** [comment]
More ubuntu madness? Honestly, why is that file even there is the linker cannot handle it.

@akesandgren There is a good reason why we use RPATH for system paths. It's the only way to make sure that those binaries use the system compiler & libs instead of the EasyBuild created ones. The binutils binaries on Fedora also have a RPATH set to the system path.

What we can do is also add our flex library to RPATH. But even better would be to statically link it in.

**Comment by akesandgren** [comment]
There is no Ubuntu madness here, this applies to any distribution which could potentially do the same thing.

The linker handles it correctly, by putting in the required objects from the static libfl_pic.a into the created binary (or shared lib if used during gcc -shared) thus resulting in a binary that does NOT have a runtime dependency on libfl.so.

The problem is that ar and ranlib from the easybuilt (with toolchain != dummy) binutils has a dependency on flex, but since RPATH is set with /usr/lib etc first it picks up the libfl.so from the system instead of the easybuilt libfl.so which comes after /usr/... in the RPATH. This is plain wrong!

The binutils built with the dummy toolchain does not need /usr/lib etc in RPATH at all, since all it depends on is
libbfd-2.26.20160125 which it must (and will) pick up from the dummy-toolchain built binutils
and
libc.so.6 which must always come from the system anyway.

Any non-dummy-toolchain binutils should never use anything from the system distribution except libc, so it should therefore NOT have /usr/lib, etc added to RPATH.

So, in conclusion, setting RPATH to include /usr/lib etc is ALWAYS wrong.

**Comment by wpoely86** [comment]
Wait, about which binutils are we talking here? The one used to bootstrap GCCcore should have the RPATH to the systems paths but the one build on top of GCCcore shouldn't have one.

As said before, there is a very good reason why we RPATH the system paths: `libstdc++.so.6`. It has been broken in the past. We need to make 100% sure that it only uses system libraries.

**Comment by akesandgren** [comment]
(Talking about binutils 2.26 only here, haven't checked any older ones, since they should basically die anyway)

There is no distinction in the easyblock for binutils whether it is the one for bootstrapping GCCcore (i.e. the one built with the dummy toolchain) or not, so all versions of binutils gets /usr/... added to the front of RPATH. None of them needed it.

The only dependencies in any of the binaries or libs in the dummy-toolchain built binutils are
libc, libdl, and libm (which always comes from the system)
and
libbfd and libopcodes (which comes from binutils itself)

There is no dependency on libstdc++ in there at all.

So, at the very least the easyblock needs to be fixed so that it only adds /usr/lib etc, to RPATH when it is built with the dummy-toolchain.

**Comment by wpoely86** [comment]
That's weird. The binutils we build with EasyBuild do link to libstdc++. You're saying it doesn't happen with the binutils you build with EasyBuild?

I try to find some time to look why that happens.

**Comment by wpoely86** [comment]
I did a quick check and as I thought, gold is written in C++. So I'm expecting to see linkage to libstdc++ there. Can you check it's really not there with the binutils build by EasyBuild on your system?

**Comment by akesandgren** [comment]
b-an01 [bin]$ ldd ld.gold 
	linux-vdso.so.1 =>  (0x00007fffd0121000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f9c4cec3000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f9c4cbba000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f9c4c7f0000)
	/lib64/ld-linux-x86-64.so.2 (0x00005589cc6cb000)

**Comment by wpoely86** [comment]
@akesandgren can you share your build log of binutils?

**Comment by akesandgren** [comment]
Want me to attach them in here or send by mail?

**Comment by wpoely86** [comment]
Mail or use github gists.

**Comment by akesandgren** [comment]
Mailed to your gmail account.

**Comment by akesandgren** [comment]
Another thing, the GCCcore-... built binutils not only have a builddependency on flex but a runtime dependency on it, since the binaries are dynamically linked with libfl.so.

The easyconfig doesn't reflect that. Changing that is probably enough, or link static against flex.

**Comment by boegel** [comment]
@akesandgren maybe you have `libstdc++.a` statically linked in your build (while we don't, because it's not there on our systems)?

I'm happy to revise the RPATH to `/usr/lib/` and co, but we did (think) we had a good reason to do so (although we may indeed be doing it more often than we should be).

I think @ocaisa was also involved with that discussion.

It seems to me that the real problem is that we're not guaranteed to statically link with flex, which would resolve the problem at hand at least (which does not mean we shouldn't correct the RPATH'ing if it's now done too often).

**Comment by boegel** [comment]
@akesandgren this should be fixed by the changes in https://github.com/hpcugent/easybuild-easyblocks/pull/1026, let us know if only RPATH'ing `/usr/lib*` when building `binutils` with `dummy` is insufficient to fix the problems you were seeing.

Thanks for reporting this!

**Comment by bigmoe** [comment]
Still seeing this issue in EasyBuild 3.0.1 on Debian sid. /usr/lib is apparently still included in rpath. 

Is there any way to manually link the static PIC Flex libraries while building binutils with dummy via EasyConfig?  Or at least force the use of the Flex libraries built with dummy? 

I've also tried to --rpath-filter /usr/lib but that didn't work. I'm new to EasyBuild though so I might be doing it wrong.

**Comment by boegel** [comment]
I'm not sure if copying from the OS is a good idea...

@bigmoe Can you provide a (debug) log for this for us to look into?

**Comment by andreas-h** [comment]
I'm now also running into this issue on my Debian/Stretch laptop. When installing GCCcore/6.3.0  (as  dependency of foss/2017a), I  get  this

    ar: error while loading shared libraries: /usr/lib/x86_64-linux-gnu/libfl.so.2: invalid ELF header
    Makefile:249: recipe for target 'libiberty.a' failed

Debug log is at https://gist.github.com/andreas-h/7e966b3e17a74e404583dd1373cb152f

**Comment by jrbosch** [comment]
I have similar problem installing binutils-2.26 dummy toolchain. 
OS. Debian 8.
In sanity check step fail.
I attach the end of the log.
[log.txt](https://github.com/hpcugent/easybuild-easyconfigs/files/977223/log.txt)

**Comment by diehlsa** [comment]
I have the same error as @jrbosch on debian 8 for binutils-2.27-dummy:

```
== FAILED: Installation ended unsuccessfully (build directory: /dev/shm/binutils/2.27/dummy-): build failed (first 300 chars): cmd "ldd /home/users/sdiehl/.local/resif/devel/default/software/tools/binutils/2.27/bin/ar" exited with exitcode 1 and output:
/home/users/sdiehl/.local/resif/devel/default/software/tools/binutils/2.27/bin/ar: error while loading shared libraries: /usr/lib64/libfl.so.2: invalid ELF header
```

**Comment by jrbosch** [comment]
@diehlsa libfl.so comes from libfl-dev (which is a dependency of flex). So when you uninstall the package, installing binutils works fine. You can check in easybuild mailing list the answers that other easybuilders gave me, searching "binutils fails because libfl.so.2".

**Comment by branfosj** [comment]
As this bug is in a now archived easyconfig I'll close the issue. If this is a problem with a newer version then please do let us know.
